File: python/drepr/models/preprocessing.py
from dataclasses import dataclass
from enum import Enum
from typing import Optional, Union

from .path import Path


@dataclass
class PMap:
    resource_id: str
    path: Path
    code: str
    output: Optional[str] = None
    change_structure: Optional[bool] = None


@dataclass
class PFilter:
    resource_id: str
    path: Path
    code: str
    output: Optional[str] = None


@dataclass
class PSplit:
    resource_id: str
    path: Path
    code: str
    output: Optional[str] = None


class RMapFunc(Enum):
    Dict2Items = "dict2items"


@dataclass
class RMap:
    resource_id: str
    path: Path
    func_id: RMapFunc
    output: Optional[str] = None


class PreprocessingType(Enum):
    pmap = "pmap"
    pfilter = "pfilter"
    psplit = "psplit"
    rmap = "rmap"


@dataclass
class Preprocessing:
    type: PreprocessingType
    value: Union[PMap, PFilter, RMap]

    @staticmethod
    def deserialize(raw: dict):
        type = PreprocessingType(raw['type'])
        raw['value']['path'] = Path.deserialize(raw['value']['path'])
        if type == PreprocessingType.pmap:
            value = PMap(**raw['value'])
        elif type == PreprocessingType.pfilter:
            value = PFilter(**raw['value'])
        elif type == PreprocessingType.psplit:
            value = PSplit(**raw['value'])
        elif type == PreprocessingType.rmap:
            value = RMap(**raw['value'])
        else:
            raise NotImplementedError()

        return Preprocessing(type, value)

File: src/lang/preprocessing.rs
use readers::path_expr::PathExpr;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(tag = "type")]
pub enum PreprocessingFunc {
  #[serde(rename = "pmap")]
  PyMap(PyMap),
  #[serde(rename = "pfilter")]
  PyFilter(PyFilter),
  #[serde(rename = "psplit")]
  PySplit(PySplit),
  #[serde(rename = "rmap")]
  RuMap(RuMap),
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct PyMap {
  pub resource_id: usize,
  pub path: PathExpr,
  pub output: Option<usize>,
  pub change_structure: Option<bool>,
  pub code: String,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct PyFilter {
  pub resource_id: usize,
  pub path: PathExpr,
  pub output: Option<usize>,
  pub code: String,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct PySplit {
  pub resource_id: usize,
  pub path: PathExpr,
  pub output: Option<usize>,
  pub code: String,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct RuMap {
  pub resource_id: usize,
  pub path: PathExpr,
  pub func_id: BuiltinRustMapFunc,
  pub output: Option<usize>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(tag = "t")]
pub enum BuiltinRustMapFunc {
  #[serde(rename = "dict2items")]
  Dict2Items
}